Certifier OSV: fixed emit func when polling

mrizzi opened the pull request:
# Description of the PR

Fixes #1387 

New Goroutine to "flush" the documents collected in the `totalDocs` based upon one condition being true:

- a 30 seconds ticker
- 1000 documents appended to `totalDocs` 

This also fixes the other bug, i.e. not exiting with `CTRL-C` 

# PR Checklist

- [X] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [X] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

pxp928 commented on cmd/guacone/cmd/osv.go:
```diff
@@ -84,11 +84,55 @@ var osvCmd = &cobra.Command{
+			for len(docChan) > 0 {
+				totalNum += 1
+				totalDocs = append(totalDocs, <-docChan)
+				if len(totalDocs) >= 1000 {
```
if there are any documents left, we should `MergedIngest` them. To capture any stragglers

mrizzi replied:
good catch, fixed

naveensrinivasan commented on cmd/guacone/cmd/osv.go:
```diff
@@ -84,11 +84,62 @@ var osvCmd = &cobra.Command{
 		gotErr := false
+
+		go func() {
+			for !gotErr {
```
The `gotErr` is shared and isn't thread-safe.

naveensrinivasan commented on cmd/guacone/cmd/osv.go:
```diff
@@ -84,11 +84,62 @@ var osvCmd = &cobra.Command{
+				select {
+				case <-ticker.C:
+					if len(totalDocs) > 0 {
+						err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
```
The `totalDocs` slice could potentially run into a race condition.

The totalDocs is accessed and modified in the goroutine that handles document ingestion. However, it's also accessed and potentially modified after the wg.Wait() call when the context error is checked and possibly when final ingestion is performed.

If the goroutine is still running and modifying totalDocs when the main goroutine reaches the final ingestion step, a race condition could occur. This is because both goroutines could be trying to access and modify totalDocs at the same time.

mrizzi replied:
@naveensrinivasan thanks a lot, it helped me a lot.
I've refactored the whole execution and now `totalDocs` slice is referenced only within the `ingestion` func.

pxp928 replied:
Thanks @naveensrinivasan that was a good catch.

naveensrinivasan replied:
Thanks. 

- A suggestion is to `break` instead of `flag` for breaking the loop.
- The `gotErr` can also run into a race condition.

```golang
		var gotErr int32
		var wg sync.WaitGroup
		ingestion := func() {
			defer wg.Done()
			var totalDocs []*processor.Document
			const threshold = 1000
		loop:
			for {
				select {
				case <-ticker.C:
					if len(totalDocs) > 0 {
						err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
						if err != nil {
							logger.Errorf("unable to ingest documents: %v", err)
							atomic.StoreInt32(&gotErr, 1)
							break loop
						}
						totalDocs = []*processor.Document{}
					}
					ticker.Reset(tickInterval)
				case d := <-docChan:
					totalNum += 1
					totalDocs = append(totalDocs, d)
					if len(totalDocs) >= threshold {
						err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
						if err != nil {
							atomic.StoreInt32(&gotErr, 1)
							logger.Errorf("unable to ingest documents: %v", err)
							break loop
						}
						totalDocs = []*processor.Document{}
						ticker.Reset(tickInterval)
					}
				case <-ingestionStop:
					break loop
				case <-ctx.Done():
					return
				}
			}
			for len(docChan) > 0 {
				totalNum += 1
				totalDocs = append(totalDocs, <-docChan)
				if len(totalDocs) >= threshold {
					err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
					if err != nil {
						atomic.StoreInt32(&gotErr, 1)
						logger.Errorf("unable to ingest documents: %v", err)
					}
					totalDocs = []*processor.Document{}
				}
			}
			if len(totalDocs) > 0 {
				err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
				if err != nil {
					atomic.StoreInt32(&gotErr, 1)
					logger.Errorf("unable to ingest documents: %v", err)
				}
			}
		}
```

mrizzi replied:
@naveensrinivasan thanks a lot :pray:

naveensrinivasan replied:
@mrizzi  Appreciate taking the feedback 👍

pxp928 replied:
seems like the flag is a better option for breaking to loop instead of a goto statement

mrizzi replied:
:sweat_smile: 
So, I have no strong opinion on this guys, but I'm concerned the certifier OSV polling is broken.

The only rationale to decide how to implement this that I can see is the consistency with the rest of guac. It looks to me that flag approach is widely used.

@naveensrinivasan @pxp928 WDYT?

naveensrinivasan replied:
I am good :)

pxp928 replied:
Yeah, let's stick with the flag. Thanks @mrizzi

mrizzi commented:
2efd048 :

- refactored `totalDocs` slice management that now happens only inside `ingestion` func
- `ingestionStop` chan to properly stop `ingestion` func
- removed last call to `ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)` because now all the ingestor invocations happen only within the  `ingestion` func
-  `ingestion` func deals with `wg WaitGroup`

naveensrinivasan commented on cmd/guacone/cmd/osv.go:
```diff
@@ -191,7 +192,7 @@ var osvCmd = &cobra.Command{
 		cf()
 
-		if gotErr {
+		if gotErr == 1 {
```
This has to use the `atomic` package to compare. To avoid the race. https://pkg.go.dev/sync/atomic#LoadInt32
